nntrainer/tensor/cache_pool.h

```diff
@@ -0,0 +1,185 @@
+// SPDX-License-Identifier: Apache-2.0
+/**
+ * Copyright (C) 2022 Jiho Chu <jiho.chu@samsung.com>
+ *
+ * @file   cache_pool.h
+ * @date   01 July 2022
+ * @see    https://github.com/nnstreamer/nntrainer
+ * @author Jiho Chu <jiho.chu@samsung.com>
+ * @bug    No known bugs except for NYI items
+ * @brief  Cache pool class inherited from memory pool
+ *
+ */
+
+#ifndef __CACHE_POOL_H__
+#define __CACHE_POOL_H__
+
+#include <mutex>
+#include <memory_pool.h>
+#include <list>
+#include <nntrainer_log.h>
+#include <swap_device.h>
+
+namespace nntrainer {
+
+/**
+ * @class   CacheElem
+ * @brief   Cache element containing swap address
+ */
+class CacheElem {
+public:
+  /**
+   * @brief CacheElem default constructor
+   *
+   */
+  explicit CacheElem(std::shared_ptr<SwapDevice> dev, unsigned int mem_id,
+                     int off, size_t len,
+                     std::shared_ptr<MemoryData<float>> data,
+                     std::vector<unsigned int> order) :
+    device(dev),
+    active(false),
+    id(mem_id),
+    offset(off),
+    length(len),
+    mem_data(data),
+    exe_order(order) {}
+
+  /**
+   * @brief CacheElem destructor
+   *
+   */
+  virtual ~CacheElem() {}
+
+  /**
+   * @brief load data from swap device
+   *
+   */
+  void swapIn();
+
+  /**
+   * @brief unload data to swap device
+   *
+   */
+  void swapOut();
+
+  /**
+   * @brief unload data to swap device
+   *
+   * @return active status
+''   */
+  bool isActive() const { return active; }
+
+  /**
+   * @brief get execution orders
+   *
+   * @return execution orders
+   */
+  std::vector<unsigned int> &getExeOrder() { return exe_order; }
+
+  /**
+   * @brief get length of cache element
+   *
+   * @return length of cache element in byte
+   */
+  size_t getLength() const { return length; }
+
+
+  unsigned int getId() const { return id; }
+
+private:
+  std::mutex device_mutex; /**< protect device */
+  std::shared_ptr<SwapDevice> device; /**< swap device */
+  bool active;                        /**< element is loaded */
+  unsigned int id;                    /**< memory id */
+  int offset;                         /**< element offset from swap device */
+  size_t length;                      /**< element size */
+  std::shared_ptr<MemoryData<float>> mem_data; /**< allocated memory data */
```

Comment:
It could be considered later PR, but how about enherit the MemoryData for CacheElem, rather than has it.

Reply:
The chracteristic of MemoryData and CacheElem is little different. I'll investigate how it can be inherited.